content/post/github_topic_modeling.md

```diff
@@ -0,0 +1,273 @@
+
+Normally, topic modeling is applied to ordinary text documents. We at source{d}
+struggle with huge amount of data originated from GitHub repositories (an soon
+  we'll handle every open source repository in the world). A natural idea is
+to treat every repository as a bag-of-words and apply the powerful engine of
+topic modeling of Dr. Vorontsov. For example, we could extract names from the
+source code and consider them "words". This article elaborates in this direction,
+presents the results and explains how to repeat them.
+
+![topic cloud](wordcloud.png)
+
+TL;DR:
+```
+docker run srcd/github_topics apache/spark
+```
+(replace `apache/spark` with any GitHub repository you wish to analyze).
+
+[OpenDocument spreadsheet with the extracted topics](https://storage.googleapis.com/github-repositories-topic-modeling/topics.ods).
+
+[JSON with the extracted topics](https://storage.googleapis.com/github-repositories-topic-modeling/topics.json).
+
+[Trained model](https://storage.googleapis.com/github-repositories-topic-modeling/repo_topic_modeling.pickle.gz) - 40M, gzipped pickle for Python 3.4+, Pandas 1.18+.
+
+Theory
+------
+Suppose that we have a topic probabilistic model of the collection of documents
+\\(D\\) which describes the occurrence of terms \\(w\\) in documents \\(d\\) with topics \\(t\\):
+$$
+p(w|d) = \\sum_{t\\in T} p(w|t) p(t|d)
+$$
+Here \\(p(w|t)\\) is the probability of the term \\(w\\) to belong to the topic \\(t\\),
+\\(p(t|d)\\) is the probability of the topic \\(t\\) to belong to the document \\(d\\),
+thus the whole formula is just the expression of [total probability](https://en.wikipedia.org/wiki/Law_of_total_probability).
+We'd like to solve the problem of recovering \\(p(w|t)\\) and \\(p(t|d)\\) from
+the given set of documents \\(\\left\\{d\\in D: d = \\left\\{w\_ 1 \\dots w\_ {n\_ d}\\right\\}, |d| = n\_ d\\right\\}\\). The terms
+belong to the vocabulary \\(W\\). We usually assume \\(\\hat{p}(w|d) = \\frac{n\_ {dw}}{n\_ d}\\),
+but this implies that all the terms are equally important which is not always true.
+Let us denote the recovered probabilities as \\(\\hat{p}(w|t) = \\phi\_ {wt}\\) and \\(\\hat{p}(t|d) = \\theta\_ {td}\\).
+Thus our problem is the stochastic matrix decomposition which is not correctly stated:
+$$
+\\frac{n\_ {dw}}{n\_ d} ≈ \\Phi \\cdot \\Theta = (\\Phi S)(S^{-1}\\Theta) = \\Phi' \\cdot \\Theta'
+$$
+It is said that [regularization](https://en.wikipedia.org/wiki/Regularization_(mathematics))
+is desirable for stating machine learning problems; in our case,
+this is *obligatory*. In other words, if we don't add any additional constraints,
+we are going to end up bad.
+
+The problems like the one from above can be solved by applying
+[maximum likelihood estimation](https://en.wikipedia.org/wiki/Maximum_likelihood_estimation):
+$$
+\\sum\_ {d\\in D}\\sum\_ {w\\in d}n\_ {dw}\\ln \\sum\_ {t}\\phi\_ {wt} \\theta\_ {td} \\to \\max_{\\Phi,\\Theta}
```

Comment:
It is quite hard to render this in my head, but I think

  \\sum\_ {w\\in d}

should be

  \\sum\_{w\in w\_{d}}